src/QsCompiler/Tests.Compiler/CompletionParsingTests.fs

```diff
@@ -0,0 +1,166 @@
+    test "function Foo (f : (Int -> " [Type]
+    test "function Foo (f : (Int -> Int)) : Unit" [Type]
+    test "function Foo (f : (Int -> Int)) : Unit " []
+    test "function Foo (q : ((Int -> Int) -> " [Type]
+    test "function Foo (q : ((Int -> Int) -> Int " []
+    test "function Foo (q : ((Int -> Int) -> Int)) : Unit" [Type]
+    test "function Foo (q : ((Int->Int)->Int)) : Unit" [Type]
+    test "function Foo (q : ((Int -> Int) -> Int)) : Unit" [Type]
+    test "function Foo<" [Declaration]
+    test "function Foo<'" [Declaration]
+    test "function Foo<> (q : 'T) : Unit" [Type]
+    test "function Foo<'T> (q : 'T) : Unit" [Type]
+    test "function Foo<'T> (q : ('T -> Int)) : Unit" [Type]
+    test "function Foo<'T> (q : ('T->Int)) : Unit" [Type]
+    test "function Foo<'T> (q : (MyType -> Int)) : Unit" [Type]
+
+[<Fact>]
+let ``Operation declaration parser tests`` () =
+    test "operation " [Declaration]
+    test "operation Foo" [Declaration]
+    test "operation Foo " []
+    test "operation Foo (" [Declaration]
+    test "operation Foo (q" [Declaration]
+    test "operation Foo (q :" [Type]
+    test "operation Foo (q : Qubit" [Type]
+    test "operation Foo (q : Qubit)" []
+    test "operation Foo (q : Qubit) :" [Type]
+    test "operation Foo (q : Qubit) : " [Type]
+    test "operation Foo (q : Qubit) : Unit" [Type]
+    test "operation Foo (q : Qubit) : Unit " [Keyword "is"]
+    test "operation Foo (q : Qubit) : Unit i" [Keyword "is"]
+    test "operation Foo (q : Qubit) : Unit is" [Keyword "is"]
+    test "operation Foo (q : Qubit) : Unit is " [Characteristic]
+    test "operation Foo (q : Qubit) : Unit is A" [Characteristic]
+    test "operation Foo (q : Qubit) : Unit is Adj" [Characteristic]
+    test "operation Foo (q : Qubit) : Unit is Adj " []
+    test "operation Foo (q : Qubit) : Unit is Adj +" [Characteristic]
+    test "operation Foo (q : Qubit) : Unit is Adj + " [Characteristic]
+    test "operation Foo (q : Qubit) : Unit is Adj + Ctl" [Characteristic]
+    test "operation Foo (q : Qubit) : Unit is Adj + Ctl " []
+    test "operation Foo (q : Qubit) : Unit is Adj + Cat " []
+    test "operation Foo (q : Qubit) : (Int, MyT" [Type]
+    test "operation Foo (q : Qubit) : (Int," [Type]
+    test "operation Foo (q : Qubit) : (Int, MyT " []
+    test "operation Foo (q : (" [Type]
+    test "operation Foo (q : (Qubit" [Type]
+    test "operation Foo (q : (Qubit," [Type]
+    test "operation Foo (q : (Qubit, " [Type]
+    test "operation Foo (q : (Qubit, Qubit" [Type]
+    test "operation Foo (q : (Qubit, Qubit " []
+    test "operation Foo (q : (Qubit, Qubit), " [Declaration]
+    test "operation Foo (q : (Qubit => " [Type]
+    test "operation Foo (q : (Qubit => Unit" [Type]
+    test "operation Foo (q : (Qubit => Unit " [Keyword "is"]
+    test "operation Foo (q : (Qubit => Unit is " [Characteristic]
+    test "operation Foo (q : (Qubit => Unit is Adj)) : Unit" [Type]
```

Comment:
Does     test "operation Foo (q : ((Qubit => Unit) is Adj)" []
work as well? If not, no worried - I think it's acceptable to not give completions in that case. I'm just curious.

Reply:
Yes, this works - it can parse the operation input/output type with or without parentheses.